### PR TITLE
feat: caching wasmtime engine

### PR DESCRIPTION
**Pull request by Stebalien**
At the moment, the cache is pretty dumb and doesn't let us "precompile". But it's a place to start and it solves the core performance problem.

### CODE REVIEW

File: fvm/src/machine/default.rs [hunk 1]
```diff
@@ -166,6 +164,10 @@ where
 
     #[cfg(feature = "builtin_actors")]
     fn load_module(&self, code: &Cid) -> Result<Module> {
+        if let Some(code) = self.engine.get(code) {
+            return Ok(code);
+        }
+
         use anyhow::Context;
         // TODO: cache compiled code, and modules?
```
Comment (raulk) [inline fvm/src/machine/default.rs:172]: Can probably remove this TODO.

File: fvm/src/machine/default.rs [hunk 2]
```diff
@@ -166,6 +164,10 @@ where
 
     #[cfg(feature = "builtin_actors")]
     fn load_module(&self, code: &Cid) -> Result<Module> {
+        if let Some(code) = self.engine.get(code) {
```
Comment (raulk) [inline fvm/src/machine/default.rs:167]: Comment would be great here.

File: testing/conformance/tests/runner.rs [hunk 3]
```diff
@@ -190,11 +190,13 @@ fn compare_state_roots(bs: &MemoryBlockstore, root: &Cid, vector: &MessageVector
 async fn conformance_test_runner() -> anyhow::Result<()> {
     pretty_env_logger::init();
 
+    let engine = Engine::default();
```
Comment (raulk) [inline testing/conformance/tests/runner.rs:193]: For M1, do we intend to have Lotus create a single engine for the full process lifetime?

**Review by raulk — APPROVED** [review]
LGTM as a starting point. For M1 it's probably ok for the client to instantiate a singleton Engine for the full lifetime of the process. And it's also OK for that engine to hold the compiled modules in memory, because it's only the built-in actors.

I think we should prioritise precompilation. It's a better experience if built-in actors are compiled as part of node initialization instead of on the first message that touches them.

**Comment by Stebalien** [comment]
> I think we should prioritise precompilation. It's a better experience if built-in actors are compiled as part of node initialization instead of on the first message that touches them.

Agreed.

File: testing/conformance/tests/runner.rs (hunk 3 above)
Comment (Stebalien) [reply testing/conformance/tests/runner.rs:193]: Yes, the FFI branch currently does that (stores it in a lazy static).

**Pull request merged**